HTBHF-1450 updated UpdatableClaimFields to actually perform the update

YetAnotherMatt opened the pull request:
This is for comment only, NOT TO BE MERGED

YetAnotherMatt commented on api/src/main/java/uk/gov/dhsc/htbhf/claimant/model/UpdatableClaimantFields.java:
```diff
@@ -1,16 +1,107 @@
 
-    EXPECTED_DELIVERY_DATE("expectedDeliveryDate");
+    EXPECTED_DELIVERY_DATE("expectedDeliveryDate"),
+    ADDRESS("address") {
```
I added address as it needs special handling. We'd also add other fields such as name etc

YetAnotherMatt commented on api/src/main/java/uk/gov/dhsc/htbhf/claimant/model/UpdatableClaimantFields.java:
```diff
@@ -1,16 +1,107 @@
+     */
+    UpdatableClaimantFields(String fieldName) {
+        this.fieldName = fieldName;
+        Map<String, Method> methodMap = Arrays.stream(Claimant.class.getDeclaredMethods()).collect(Collectors.toMap(
```
I tried to make this map static, but it couldn't then be referred to from an enum constructor

mikeSmith08 commented:
My thoughts on this are:
 - using reflection to build a map of all the getters and setters based on their name is quite clever, but seems overly complicated compared to the alternative which is to simply define them. I know this might become a maintenance nightmare, but is that any worse than complicated reflection code, plus you've already got that for the address anyway? Very much playing devil's advocate here, but I thought it would be good to say it anyway.
 - my main concern is that I don't like the main usage in the Production code in `NewClaimService`. I think it's a bit magical that the enum is doing the updating to the object as well as doing the comparison in one foul swoop. I think if we had something like:
   ```
if (isClaimantUpdated(originalClaimant, claimant)) {
    updateOriginalClaimant(originalClaimant, claimant);
}
```
I realise that this would need a lot of change from what you've written, but I worry that its too complicated for its own good when a simpler solution might suffice. I would love to pair through this with you and see what the simpler alternative would look like, but I realise that time is against me here so will go with whatever you decide.

YetAnotherMatt commented:
@dwybourn on Mike's suggestion I've removed the reflection and moved the control flow to NewClaimService. I think it probably is better for it. Have a look and let me know what you think, then I'll close this PR and put some tests in, and the other fields, when I get a chance.
